py-rse/style.Rmd

```diff
@@ -59,165 +48,228 @@ Look for algorithmic problems first.
+Three common patterns that people use for this are:
 
-### Keep functions short.
+1.  Figure out what the user has asked it to do (Chapter \@ref(py-rse-configuration)).
+2.  Read all input data.
+3.  Process it.
+4.  Write output.
 
-Nothing should be more than a page long
-or have more than three levels of indentation because of nested loops and conditionals.
-Anything longer or more deeply nested will be hard for readers to fit into [working memory][working-memory];
-when you find yourself needing to break these limits,
-extract a function.
+or:
 
-### Handle special cases at the start of the function.
+1.  Figure out what the user has asked for.
+2.  For each input file:
+    1.  Read.
+    2.  Process.
+    3.  Write file-specific output (if any).
+3.  Write summary output (if any).
 
-This helps readers mentally get them out of the way
-and focus on the "normal" case.
+or:
+
+1.  Figure out what the user has asked for.
+2.  Repeatedly:
+    1.  Wait for user input.
+    2.  Do what the user has asked.
+3.  Exit when a "stop" command of some sort is received.
+
+Each step in each of the outlines above usually becomes a function.
+Those functions depend on others,
+some of which are written to break code into comprehensible chunks and are then called just once,
+others of which are utilities that may be called many times from many different places.
+We put all of the single-use functions in the first half of the file
+in the order in which they are likely to be called,
+and then put all of the multi-use utility functions in the bottom of the file in alphabetical order.
+If any of those utility functions are used by other scripts or programs,
+they should go in a file of their own.
+In fact,
+this is a good practice even if they're only used by one program,
+since it signals even more clearly which functions are specific to this program
+and which are likely to be reused elsewhere.
 
 ## How can I check that code follows style guidelines? {#py-rse-style-check}
 
-Checking that code conforms to guidelines like the ones above can be time consuming,
-but luckily it doesn't have to be done by hand.
-Most languages have tools that will check code style rules and report violations.
-These are often called [linters][linter],
+Checking that code conforms to guidelines like PEP-8 would be time consuming
+if it had to be done manually,
+but luckily it doesn't.
+Most languages have tools that will check style rules;
+these are often called [linters][linter],
 after an early tool called `[lint][lint]` that found lint (or fluff) in C code.
-Python has a tool that used to be called `pep8`
-and is now called `pycodestyle`
-that will do this.
-For example,
-this program is supposed to count the number of stop words in a document:
+
+Python's linter used to be called `pep8` and is now called `pycodestyle`.
+To see how it works,
+let's look at this program,
+which is supposed to count the number of [stop words][stop-word] in a document:
```

Comment:
Instead of defining a new program, perhaps we should run the linter over one of the shorter scripts we've developed (`countwords.py` or `collate.py`)?

Reply:
The downside of course is that we'd need to write sub-optimal code earlier on in order for the linter to pick something up...

Reply:
That was my thinking...